changelog/dmd.source-archive.dd

```diff
@@ -0,0 +1,97 @@
+an offset/length to the filename string, and an offset/length to the file
+contents
+
+3. the filename strings, each string has a terminating 0
+
+4. the file contents, each file has four 0 bytes appended, as the lexer wants
+that as a sentinel
+
+5. the integers in the format are little-endian
+
+To create a .sar file, such as one for Phobos' std:
+
+dmd -sar=/home/duser/dmd/src/phobos/std
+
+and the file:
+
+/home/duser/dmd/src/phobos/std.sar
+
+will be created and filled with all sub-files with one of the extensions ".di", ".d",
+".c", or ".i". The filenames in the .sar file will include the path prefix
+specified in the -sar switch. This means the .sar file cannot be successfully moved
+to another location.
+
+For Phobos, std.sar is approximately 11 megabytes in size.
+
+To use the std.sar file, nothing needs to be changed in the user's build system.
+dmd will automatically prefer using any .sar file it finds. To disable using
+.sar files, which would be necessary when doing development of the source files,
+select one of the following:
+
+1. delete the .sar file
+
+2. use the -sar=off compiler switch. -sar=on turns it on, and is the default
+setting
+
+Trying out .sar with simple programs like hello world yield a negligible difference
+in compile speed. It's unlikely a larger program would show any particular trend
+in performance.
+
+A standalone archiver program can be easily created from the implementation in DMD.
+
+## Rationale
+
+1. All the source files in a project or library can be represented as a single file,
+making it easy to deal with.
+
+2. To compile all the source files at once with DMD, the command line can get
+extremely long, and certainly unwieldy. With .sar files, you may not even need
+a makefile or builder, just:
+
+dmd project.sar
+
+3. In Phobos (and most code), the tendency is to lump a lot of only marginally related
+functions into one file. This is likely because of the inconvenience of multiple files.
+In std.algorithm, for example, the individual algorithms could be placed into multiple
+files, since they don't refer to each other. This could also help the people who
+don't want the automatic "friend" status of declarations within a single module.
+.sar files can make much more granular modules more attractive.
+
+4. A directory in a project can be anything. But multiple .sar files in a project means
+that's where the code is. Multiple versions of the project can exist in the same directory
+by using .sar files with different names.
+
+5. Experiments with small programs (hello world) show a negligible change in compile
+speed. Much larger programs may show significant compile speed increases with .sar
+files, simply because there are fewer file operations. For slow file systems, such as
+SD cards, or network file systems, the speedup could be substantial.
+
+None of these make it a slam-dunk, after all, no other compiler does this that I'm
+aware of. Even so, some surprising uses can be expected of .sar files.
```

Comment:
I think Go, Rust, and C++ (cxx98 templates at a stretch, but also cxx20 modules) could be named as compilers with some sort of packaging format for speed/caching.

Reply:
- [C++-98's -frepo](https://gcc.gnu.org/onlinedocs/gcc-9.5.0/gcc/Template-Instantiation.html#Template-Instantiation) produced an .rpo for all exported template in a given build, imported for linking.
- [C++-20's modules](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Compiled-Module-Interface.html)  generate .gcm for compiled module interfaces (can contain multiple modules within IIUC) for importing into compilation on `import`.
- [Go exports](https://gcc.gnu.org/onlinedocs/gccgo/Import-and-Export.html) types and definitions to a .gox file for importing on `import`.
- [Rust crates export](https://rustc-dev-guide.rust-lang.org/backend/libs-and-metadata.html#metadata) metadata and optionally encoded MIR to a .rmeta (or .rox for gccrs) on `use`.
- I've just remembered [C has precompiled headers](https://gcc.gnu.org/onlinedocs/gcc/Precompiled-Headers.html).

Reply:
To the point, none of those are full source archives.

Reply:
In that vein, you could include the precompiled headers generated and used by my Symantec C++ compiler. They did substantially improve compile times. But it's not a system I wished to propagate to D. C++ made a mistake going with their binary import system.

Reply:
> all of them (except frepo) can/are used to replace source files for speeding up compilation.

This doesn't replace source files, which is quite an important difference e.g. lets say file system accesses are just so catastrophically expensive that this is worthwhile, it'd still be parsing them and running semantic analysis every single time (long story short this is where the actual pain is - and more importantly the memory consumption. I'd take a 10% longer build if it didn't use 70GB of ram to do a naive build of our code)

Reply:
Isn't a JAR file exactly this (and more) ?